src/PhpSpec/Runner/Maintainer/CollaboratorsMaintainer.php

```diff
@@ -143,7 +143,7 @@ private function generateCollaborators(CollaboratorManager $collaborators, \Refl
 
     private function isUnsupportedTypeHinting(\ReflectionParameter $parameter): bool
     {
-        return $parameter->isArray() || $parameter->isCallable();
+        return $parameter->getType() && in_array($parameter->getType()->getName(), ['array', 'callable']);
```

Comment:
This isn't the same? `isCallable` would return true if the type name is `Closure` too.

Reply:
Also, I think this is only deprecated in PHP 8. If you are looking for proper PHP 8 support then you will also need to check that `getType` returns an object which has `getName` (`instance of ReflectionNamedType`). Union typed objects only have a `getNames` method.

Reply:
https://github.com/laravel/framework/blob/v7.17.2/src/Illuminate/Support/Reflector.php#L10-L31 might be a helpful example RE the getName stuff.

Reply:
> This isn't the same? isCallable would return true if the type name is Closure too.

There could also be a typed class that is invokable. You'd need to check that too (and also the special case of `self`, which doesn't get resolved to the real class name).

Reply:
IIRC, `isCallable` for parameter is only about the `callable` typehint

Reply:
Oh, so not the same as `is_callable`?

Reply:
@GrahamCampbell thanks, we'll need to consider how union types will even work I guess!

Reply:
@GrahamCampbell No, `isCallable` is just about `callable`. To evaluate whether a string or array or object was callable would require autoloading and all sorts of runtime stuff

https://3v4l.org/oktBd

Reply:
Yep. You are right. That line looks great. 👍 

NB: I'm unable to mark this comment as resolved for some reason.